### ISSUE TITLE
Compile Rust to JavaScript with emscripten

### ISSUE THREAD

**Issue by brson** [Mozilla-Student-Projects/Projects-Tracker#33]
It would be nice to be able to run Rust code in the browser, particularly so we can create a 'try rust' website. We would like to do this using the [emscripten](https://github.com/kripken/emscripten) LLVM-to-JavaScript compiler. The best approach is not exactly clear, but will probably entail porting more C++ runtime code to Rust and creating some sort of platform abstraction layer.

See also: [2235](https://github.com/mozilla/rust/issues/2235)

**Comment by z0w0** [comment]
It should be as "easy" as creating an intermediary library in JS for all runtime dependencies (libuv, etc.) like [this](https://raw.github.com/kripken/emscripten/master/src/library_gc.js). I say easy in the sense that I don't think you'll have to do much magic to get it working, just a lot of work to get the runtime wrapped with JS functionality.

**Comment by jsyeo** [comment]
Hi, I will like to work on this. Are there any docs that I should consult before starting?

I have looked at https://github.com/kripken/emscripten/wiki and http://mozakai.blogspot.sg/2012/03/howto-port-cc-library-to-javascript.html.

Also, I've tried to compile it with emcc but I'm getting some errors. Should I post them to the emscripten mailing list?

**Comment by z0w0** [comment]
You will need to look into how the runtime works, what sort of memory requirements it needs and how those memory requirements would best be wrapped in Javascript.

Assuming the error you get is something about braces, then I get that too. And yeah, I would post it to the mailing list, along with the bitcode file (`rustc --emit-llvm`) and the LLVM assembly file (`llvm-dis file.bc`). Or create an issue on emscripten's github repository.

**Comment by brson** [comment]
@jsyeo by reading the emscripten docs you are a step ahead of me. You are fortunate that emscripten's author (@kripken / azakai) is on the Mozilla research team (and sits next to me) so you can feel free to bug him _extensively_. For things that you suspect are emscripten bugs, the best place to post them is the emscripten bug tracker on github, otherwise the emscripten mailing list is appropriate. #emscripten on irc.mozilla.org is a good place to ask questions too (as well as #rust).

**Comment by minhnhdo** [comment]
Any update on this?

I have tried with the following programs and they work fine.

``` rust
#[start]
fn main(_:int, _: **u8) -> int { 0 }
```

``` rust
#[start]
fn main(_: int, _: **u8) -> int {
    unsafe {
        std::libc::puts(0 as *i8);
    }
    0
}
```

**Comment by z0w0** [comment]
I'd imagine with the work towards runtime-less rust this should be much easier to implement now. And by working, do you mean those programs compile fine in Emscripten?

**Comment by minhnhdo** [comment]
Yes, they do compile and run fine.

**Comment by brson** [comment]
@mrordinaire I'd be very interested to see a blog post or other writeup about how you compiled Rust with emscripten. I've never heard of anybody doing that. The ultimate goal of this task would be to make emscripten a first-class platform in Rust so that you could e.g. `rustc --target=i686-pc-emscripten` (or whatever emcc is using as a triple), and that includes having a working standard library.

**Comment by minhnhdo** [comment]
@brson Yes, I will make a writeup of this when I can find the time (it's right in the middle of exams here on my side). About emscripten being a first class platform, how will `rustc --target=i686-pc-emscripten`'s output different from the other targets? Does that mean emscripten will be bundled with rustc?

**Comment by brson** [comment]
@mrordinaire rustc would probably need to take an argument specifically to provide the path to the emscripten toolchain, similar to our `--android-cross-path` argument.

**Comment by brson** [comment]
Still a great project. Should be pretty easy these days since the libs are very modular now.

**Comment by Arcnor** [comment]
Is there any progress with this? Do we need to modify Rust itself to make this work?

I was thinking the same, some sort of "Try Rust", as documentation is great but still no "user friendly"

**Comment by jdm** [comment]
See existing progress at http://www.reddit.com/r/rust/comments/20nnkk/rust_and_emscripten_a_small_success/

**Comment by Arcnor** [comment]
Awesome! So sad that the `std` is not ported yet. I also found this: http://bilalhusain.com/rust-lexer/inference.html, not sure if that's using ECMAScripten to do whatever is doing

**Comment by brson** [comment]
@Arcnor I just put a status update about how I think to proceed over on the [main Rust issue tracker](https://github.com/mozilla/rust/issues/2235#issuecomment-45423901). You are very welcome to take a stab at it.

**Comment by Arcnor** [comment]
@brson Just was looking at it, thank you very much :D. I'm still a super newbie at Rust, and that's why I was looking at some "to the web" version, to help or document my journey.

If I get any of your steps done, I'll comment in that issue. Thanks!

**Comment by jonnor** [comment]
It looks like https://github.com/rust-lang/rfcs/issues/604 is the most recent place for info about this
